examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+       changed, these may override an existing charging command */
+    switch (mHwState)
+    {
+    case StateEnum::kNotPluggedIn:
+        SetState(StateEnum::kNotPluggedIn);
+        break;
+
+    case StateEnum::kPluggedInNoDemand:
+        SetState(StateEnum::kPluggedInNoDemand);
+        break;
+
+    case StateEnum::kPluggedInDemand:
+        /* If the EVSE is asking for demand then enable charging */
+        SetState(StateEnum::kPluggedInCharging);
+        break;
+
+    default:
+        ChipLogError(AppServer, "Unexpected EVSE hardware state\n");
+        SetState(StateEnum::kFault);
+        break;
+    }
+
+    /* update SupplyState */
+    SetSupplyState(SupplyStateEnum::kChargingEnabled);
+
+    /* If it looks ok, store the min & max charging current */
+    mMaximumChargingCurrentLimitFromCommand = maximumChargeCurrent;
+    SetMinimumChargeCurrent(minimumChargeCurrent);
+    // TODO persist these to KVS
+
+    // TODO: Generate events
+
+    return this->ComputeMaxChargeCurrentLimit();
+}
+
+/**
+ * @brief   Called when EVSE cluster receives EnableDischarging command
+ *
+ * @param dischargingEnabledUntil (can be null to indefinite discharging)
+ * @param maximumChargeCurrent (in mA)
+ */
+Status EnergyEvseDelegate::EnableDischarging(const DataModel::Nullable<uint32_t> & dischargingEnabledUntil,
+                                             const int64_t & maximumDischargeCurrent)
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::EnableDischarging() called.");
+
+    /* update SupplyState */
+    SetSupplyState(SupplyStateEnum::kDischargingEnabled);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called when EVSE cluster receives StartDiagnostics command
+ */
+Status EnergyEvseDelegate::StartDiagnostics()
+{
+    /* For EVSE manufacturers to customize */
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::StartDiagnostics()");
+
+    /* update SupplyState */
+    SetSupplyState(SupplyStateEnum::kDisabledDiagnostics);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/* ---------------------------------------------------------------------------
+ * FUNCTIONS BELOW:
+ *    - EVSE Hardware interface
+ *
+ *  RegisterEvseHardwareCallback( callbackType, callbackFnc )
+ *  SetMaxHardwareCurrentLimit( currentmA )
+ *  SetCircuitCapacity( currentmA )
+ *  SetCableAssemblyLimit( currentmA )
+ *  SetState( EVSEStateEnum )
+ *  SetFault
+ *
+ */
+
+/**
+ * @brief    Called by EVSE Hardware to register a callback
+ *
+ * @param    Callback function
+ */
+#if 0
+Status EnergyEvseDelegate::HwRegisterEvseHardwareCallback(int Callback) // TODO
+{
+    // TODO
+    return CHIP_NO_ERROR;
+}
+#endif
+/**
+ * @brief    Called by EVSE Hardware to notify the delegate of the maximum
+ *           current limit supported by the hardware.
+ *
+ *           This is normally called at start-up.
+ *
+ * @param    currentmA
```

Comment:
delete non-useful comments. Please apply throughout.

Reply:
Reviewed them to make them more useful and fixed some incorrect comments.